### PR TITLE
[Bulk Load] Hybrid Events Bulk Loader

### PR DESCRIPTION
**Pull request by amnn**
## Description

Applying learnings from the hybrid transaction schema to the events schema:

- Partition by transaction sequence number (therefore remove checkpoint sequence number from the table and key).

- Add separate tables to do each individual filter, so that each individual filter can be implemented using a single index only scan that returns a sorted sequence of event sequence number keys, including breaking up module and type filters into separate, cascading filters.

- Add a `sender` field to each filter table so we can do combination filters with `sender`.

- Revert to treating each transaction as having one distinguished `sender`.

## Test plan

Used to generate `amnn_0_hybrid_ev*` on the read replica.

## Stack

- #19 
- #20 
- #21 
- #22

### CODE REVIEW

File: migration_scripts/bulk_load/src/hybrid_ev.clj [hunk 1]
```diff
@@ -0,0 +1,587 @@
...
+    "CREATE INDEX %1$s_sender ON %s (
+         sender,
+         package,
+         module,
+         name,
+         tx_sequence_number,
+         event_sequence_number
+     )"))
+
+;; Table: hybrid_ev_event_type ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
```
Comment (amnn) [inline migration_scripts/bulk_load/src/hybrid_ev.clj:257]: In this table `name` is the simple type name including all type parameters, e.g. `CoinMetadata<0x0...0::sui::SUI>`

File: migration_scripts/bulk_load/src/hybrid_ev.clj [hunk 2]
```diff
@@ -0,0 +1,587 @@
...
+  (db/with-table! db +ev-event-mod+
+    "CREATE INDEX %1$s_sender ON %s (
+         sender,
+         package,
+         module,
+         tx_sequence_number,
+         event_sequence_number
+     )"))
+
+;; Table: hybrid_ev_event_name ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
```
Comment (amnn) [inline migration_scripts/bulk_load/src/hybrid_ev.clj:218]: In this table, `name` is just the type name, without type parameters, e.g. `CoinMetadata`.

File: migration_scripts/bulk_load/src/hybrid_ev.clj [hunk 3]
```diff
@@ -0,0 +1,587 @@
...
+         event_sequence_number
+     )"))
+
+;; Table: hybrid_ev_event_pkg ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn ev-event-pkg:create! [db]
+  (jdbc/with-transaction [tx db]
+    (db/with-table! db +ev-event-pkg+
+      "CREATE TABLE %s (
+           package                     BYTEA,
```
Comment (emmazzz) [inline migration_scripts/bulk_load/src/hybrid_ev.clj:155]: Double checking that this is the package of the event struct type itself, not the package in which the event is emitted right?

Comment (emmazzz) [reply migration_scripts/bulk_load/src/hybrid_ev.clj:155]: Ah I see the insertion statements below

Comment (amnn) [reply migration_scripts/bulk_load/src/hybrid_ev.clj:155]: That's right -- there's two tables: `ev-emit-pkg` and `ev-event-pkg`. The former is the package containing the function called in the PTB that generated the event, and the latter (this one) is the package that introduced the event type.

**Comment by amnn** [comment]
## Size analysis

The original schema weighs in at **683GiB**, and the proposed hybrid schema weights **1529GiB** for a 51 day period (so just over twice as big), which is not great. But I also estimated the benefits of various optimisations on this table:

- If we are able to use clustered tables, we could save 391GiB by combining the storage for the table and its primary key (estimated by calculating the storage spent on primary keys).
- If we replace addresses and object IDs by a 64bit identifier, we can save an additional 380GiB: 24B saved per address or ID, multiplied by 
  - 4 for each table that mentions a package and a sender address: both need storage in the table and `sender` index, (the package also needs storage in the primary key, but assuming we're using clustered tables this goes away).
  - 2 for the `ev_senders` table: once in the table and once in the primary key.

So with the optimisations in place, we can get the size down to **758GiB** (roughly half the current size, and 10% over the existing schema footprint).

```
defaultdb=> SELECT pg_size_pretty(SUM(pg_relation_size(CAST(relname AS VARCHAR)))) FROM pg_class WHERE relname LIKE 'amnn_0_orig_ev%' AND relkind IN ('r', 'i', 'p');
 pg_size_pretty
----------------
 683 GB
(1 row)

defaultdb=> SELECT pg_size_pretty(SUM(pg_relation_size(CAST(relname AS VARCHAR)))) FROM pg_class WHERE relname LIKE 'amnn_0_hybrid_ev%' AND relkind IN ('r', 'i', 'p');
 pg_size_pretty
----------------
 1529 GB
(1 row)

defaultdb=> SELECT pg_size_pretty(SUM(pg_relation_size(CAST(relname AS VARCHAR)))) FROM pg_class WHERE relname LIKE 'amnn_0_hybrid_ev%_pkey' AND relkind IN ('r', 'i', 'p');
 pg_size_pretty
----------------
 391 GB
(1 row)
```

**Review by wlmyng — APPROVED** [review]

**Pull request merged**